extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java

```diff
@@ -830,4 +884,559 @@ private Table readTableInternal(
+                    verifyOverwriteCompatibility(icebergTable.schema(), newSpecAndSchema.schema);
+                    verifyOverwriteCompatibility(icebergTable.spec(), newSpecAndSchema.partitionSpec);
+                } else {
+                    verifyAppendCompatibility(icebergTable.schema(), useDefinition);
+                    verifyAppendCompatibility(icebergTable.spec(), useDefinition);
+                }
+            }
+        } else if (writeInstructions.createTableIfNotExist()) {
+            newNamespaceCreated = createNamespaceIfNotExists(tableIdentifier.namespace());
+            newSpecAndSchema = fromTableDefinition(useDefinition, userInstructions);
+            icebergTable = createNewIcebergTable(tableIdentifier, newSpecAndSchema, writeInstructions);
+            newTableCreated = true;
+        } else {
+            throw new IllegalArgumentException(
+                    "Table does not exist: " + tableIdentifier + ", update the instructions " +
+                            "to create the table if it does not exist and try again.");
+        }
+
+        try {
+            final List<CompletedParquetWrite> parquetFileinfo =
+                    writeParquet(icebergTable, dhTables, writeInstructions);
+            final List<DataFile> appendFiles = dataFilesFromParquet(parquetFileinfo);
+            if (addSnapshot) {
+                commit(icebergTable, newSpecAndSchema, appendFiles, writeInstructions, overwrite && !newTableCreated);
+            }
+            return appendFiles;
+        } catch (final RuntimeException writeException) {
+            if (newTableCreated) {
+                // Delete it to avoid leaving a partial table in the catalog
+                try {
+                    catalog.dropTable(tableIdentifier, true);
+                } catch (final RuntimeException dropException) {
+                    writeException.addSuppressed(dropException);
+                }
+            }
+            if (newNamespaceCreated) {
+                // Delete it to avoid leaving a partial namespace in the catalog
+                dropNamespaceIfExists(tableIdentifier.namespace());
+            }
+            throw writeException;
+        }
+    }
+
+    /**
+     * Check if the schema for the iceberg table is compatible for overwriting with a deephaven table with provided
+     * definition.
+     */
+    private static void verifyOverwriteCompatibility(
+            final Schema icebergSchema,
+            final Schema newSchema) {
+        if (!icebergSchema.sameSchema(newSchema)) {
+            throw new IllegalArgumentException("Schema mismatch, iceberg table schema: " + icebergSchema +
+                    ", schema derived from the table definition: " + newSchema);
+        }
+    }
+
+    /**
+     * Check if the partition spec for the iceberg table is compatible for overwriting with a deephaven table with
+     * provided definition.
+     */
+    private static void verifyOverwriteCompatibility(
```

Comment:
Should we move some of these static functions to `IcebergUtils`? Or another class for internal Iceberg functions?